### ISSUE TITLE
Kubernetes cannot apply NetworkPolicy on RHEL8 due to different source IP behavior

### ISSUE THREAD

**Issue by rayburgemeestre** [projectcalico/calico#6235]
Hi guys,

I have a pretty simple test setup on two pretty much identical Kubernetes `1.21.4` clusters; one Ubuntu 20.04 and one RHEL 8.5. I have two Pods, a webserver on `node001` on port `18081`, and a webserver-client on `node002` that periodically tries to `wget` that webserver. This works. So Kubernetes + Calico in principle seems to work fine for both clusters.

Then, I introduce a `NetworkPolicy` that explicitly allows traffic to port `18081` based on Pod labels (`YAML` after this paragraph). This should prevent any other Pods that do not match the criteria of labels in the policy to be blocked. This works correctly on Ubuntu 20.04, but _not_ on RHEL (tested 8.4 and 8.5). On RHEL it actually results in no traffic being allowed _at all_ to this Pod, even if it has matching labels.

```
kind: NetworkPolicy
apiVersion: networking.k8s.io/v1
metadata:
  name: webserver-allow-port-18081
spec:
  podSelector:
    matchLabels:
      app: webserver
  ingress:
  - ports:
    - port: 18081
    from:
    - podSelector:
        matchLabels:
          app: webserver-client
```

Calico is configured identically on both my Ubuntu and RHEL test setups. Some difference that I see with `tcpdump` that might be relevant is: On RHEL the packets arrive on the destination host (`node001`) with the IP calico uses for the sending node (`node002`) as the IPIP tunnel (`projectcalico.org/IPv4IPIPTunnelAddr`, which is `172.29.112.134`). 

Where on Ubuntu it seems that already at the sending node (`node002`) the Source IP is replaced with the Pod IP instead (`172.29.112.138`) before sending packets. So the source IP won't be that of the IPIP tunnel (which is `172.29.112.128` on the Ubuntu setup). I'll include the `tcpdump` output and other info below.

**The working Ubuntu 20.04 (kernel version 5.4.0-73-generic) setup**

```
root@node002:~# ip addr show dev tunl0
8: tunl0@NONE: <NOARP,UP,LOWER_UP> mtu 1480 qdisc noqueue state UNKNOWN group default qlen 1000
    link/ipip 0.0.0.0 brd 0.0.0.0
    inet 172.29.112.128/32 scope global tunl0
       valid_lft forever preferred_lft forever

root@node002:~# tcpdump -i tunl0 port 18081
11:54:07.687249 IP 172.29.112.138.58308 > 172.29.152.134.18081: Flags [S], seq 1817700180, win 64240, options [mss 1460,sackOK,TS val 706166123 ecr 0,nop,wscale 7], length 0
11:54:07.688276 IP 172.29.152.134.18081 > 172.29.112.138.58308: Flags [S.], seq 372371080, ack 1817700181, win 64260, options [mss 1440,sackOK,TS val 2464695229 ecr 706166123,nop,wscale 7], length 0

root@node001:~# tcpdump -i tunl0 port 18081
11:54:07.687821 IP 172.29.112.138.58308 > 172.29.152.134.18081: Flags [S], seq 1817700180, win 64240, options [mss 1460,sackOK,TS val 706166123 ecr 0,nop,wscale 7], length 0
11:54:07.688100 IP 172.29.152.134.18081 > 172.29.112.138.58308: Flags [S.], seq 372371080, ack 1817700181, win 64260, options [mss 1440,sackOK,TS val 2464695229 ecr 706166123,nop,wscale 7], length 0
```

In summary: I see the Pod IP here as the source IP.

**The broken RHEL 8.5 (kernel version 4.18.0-372.9.1.el8.x86_64) setup**

```
[root@node002 ~]# ip addr show dev tunl0
3: tunl0@NONE: <NOARP,UP,LOWER_UP> mtu 1480 qdisc noqueue state UNKNOWN group default qlen 1000
    link/ipip 0.0.0.0 brd 0.0.0.0
    inet 172.29.112.134/32 scope global tunl0
       valid_lft forever preferred_lft forever

[root@node002 ~]# tcpdump -i tunl0 port 18081
listening on tunl0, link-type RAW (Raw IP), capture size 262144 bytes
11:52:11.555868 IP 172.29.112.134.55032 > 172.29.152.167.18081: Flags [S], seq 777305220, win 28200, options [mss 1410,sackOK,TS val 3374593287 ecr 0,nop,wscale 7], length 0

[root@node001 ~]# tcpdump -i tunl0 port 18081
listening on tunl0, link-type RAW (Raw IP), capture size 262144 bytes
11:52:11.556545 IP 172.29.112.134.55032 > 172.29.152.167.18081: Flags [S], seq 777305220, win 28200, options [mss 1410,sackOK,TS val 3374593287 ecr 0,nop,wscale 7], length 0
```

In summary: I see the tunl0 IPv4IPIPTunnelAddr here as the source IP.

## Expected Behavior

The receiving node should be able to get the Source Pod IP from the packet somehow, so that it can lookup which Pod the traffic is coming from, and with that check if the labels from the NetworkPolicy match with the Pod labels and based on that allow or reject the traffic.

## Current Behavior

My analysis of the current behavior can be wrong, since I don't know if I fully understand it. I captured pcap files and tried to analyse them in wireshark: [pcaps.tar.gz](https://github.com/projectcalico/calico/files/8920490/pcaps.tar.gz). From this I got the impression, and I can very well be wrong, that the Pod IP is not embedded in the packet (which is how I assumed IPIP worked) for neither Ubuntu and RHEL, but simply the Pod IP is already put there before it's being sent in the case of Ubuntu. 

## Possible Solution

I hope by making this tickets I can get some guidance on where to look further.

## Your Environment

* Calico version v3.23.1
* Orchestrator version Kubernetes 1.21.4
* Operating System and version: Red Hat Enterprise Linux release 8.4 (Ootpa)
* OS Kernel: 4.18.0-305.el8.x86_64

One notable difference between the Ubuntu (working) setup and this RHEL8 environment is that RHEL8 uses `nftables` instead of `iptables`.

**Comment by caseydavenport** [comment]
@rayburgemeestre you're correct in that you should expect to see the original pod IP address show up in the un-encapsulated packet on the destination node. 

The fact that you're seeing the source node's tunnel IP seems to suggest that perhaps there is an extra MASQUERADE or SNAT that is occurring on the source node which is hiding the original source IP. 

One thing you could do is, while running traffic from the pod in question, monitor the following in iptables on the source node:

```
sudo watch iptables-save -c | grep MASQ
```
and
```
sudo watch iptables-save -c | grep SNAT
```

^ This will give you counters to see if there are any iptables rules performing MASQ or SNAT that are incrementing.

**Comment by rayburgemeestre** [comment]
Thanks Casey, for the pointers. It was nice to know I was on the right track and it helped me find the culprit! And sorry it was unrelated to calico in the end.

Turns out we were using an extra MASQUERADE rule in our kubelet services file that was placed there as a fix a long time ago in order to "allow networking from the container to the host network".

```
ExecStartPre=/sbin/iptables -P FORWARD ACCEPT
ExecStartPre=/sbin/iptables -t nat -A POSTROUTING -m addrtype ! --dst-type local ! -d 10.141.0.0/16 -j MASQUERADE
```

I can't fully recall how we came to this rule, but in hindsight I believe this wouldn't have been necessary if we had disabled the `--masquerade-all` flag for the kube-proxy (also running as a service). 

If I disable that flag on kube-proxy I can get rid of those iptables additions, and networking works correctly!

Just in case others that land on this ticket may find this related comment useful: In the meantime (before I solved this issue) I tried running calico in eBPF mode. It worked around this problem (of the erroneous MASQ rule)

I needed a pretty recent calico version (3.20.1 didn't work, so I updated to v3.23.1).

```
calicoctl patch felixconfiguration default --patch='{"spec": {"bpfKubeProxyIptablesCleanupEnabled": false}}'
calicoctl patch felixconfiguration default --patch='{"spec": {"bpfEnabled": true}}'
```

This was about a bare-metal cluster, in my case I had to keep kube-proxy running, which is why I also patched bpfKubeProxyIptablesCleanupEnabled to false. My plan is to investigate why I couldn't turn off kube-proxy fully, but at the time of writing I haven't figured that out yet.
